src/printer/layers.js

```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
     layerSpec.opacity
   );
 }
+
+/**
+ * @param {WfsLayer} layerSpec
+ * @param {FrameState} rootFrameState
+ * @return {Observable<LayerPrintStatus>}
+ */
+function createLayerWFS(layerSpec, rootFrameState) {
+  const width = rootFrameState.size[0];
+  const height = rootFrameState.size[1];
+  const context = createCanvasContext2D(width, height);
+  context.canvas.style = {};
+  let frameState;
+  let renderer;
+  let format;
+  let gmlFormats = {
+    '1.0.0': new GML2(),
+    '1.1.0': new GML3(),
+    '2.0.0': new GML32(),
+  };
+
+  if (layerSpec.format === 'gml') {
+    format = new WFS({
+      version: layerSpec.version,
+      gmlFormat: gmlFormats[layerSpec.version],
+    });
+  } else {
+    format = new GeoJSON();
+  }
+
+  let vectorSource = new VectorSource({
+    format: format,
+    loader: function (extent, resolution, projection) {
+      let proj = projection.getCode();
+      let xhr = new XMLHttpRequest();
+      let params = {
+        service: 'WFS',
+        request: 'GetFeature',
+        version: layerSpec.version ? layerSpec.version : '1.1.0',
+        typename: layerSpec.layer,
+        srsName: proj,
+        bbox: extent.join(','),
+      };
+      if (layerSpec.format !== 'gml') {
+        params['outputFormat'] = 'application/json';
+      }
+      let urlParams = Object.entries(params)
+        .map((e) => e.join('='))
+        .join('&');
+      let url = layerSpec.url + '?' + urlParams;
+      xhr.open('GET', url);
+      let onError = function () {
+        vectorSource.removeLoadedExtent(extent);
+        progress$.next([1, context.canvas, layerSpec.url]);
+        progress$.complete();
+      };
+      xhr.onerror = onError;
+      xhr.onload = function () {
+        if (xhr.status == 200) {
+          vectorSource.addFeatures(
+            vectorSource.getFormat().readFeatures(xhr.responseText)
+          );
+          renderer.prepareFrame({ ...frameState, time: Date.now() });
+          renderer.renderFrame(
+            { ...frameState, time: Date.now() },
+            context.canvas
+          );
+          progress$.next([1, context.canvas]);
+          progress$.complete();
+        } else {
+          onError();
+        }
+      };
+      xhr.send();
+    },
+    strategy: bbox,
+  });
+
+  let layer = new VectorLayer({
+    source: vectorSource,
+  });
+
+  frameState = {
```

Comment:
can this be factorized ?

Reply:
Indeed I have the impression that we only need the `layer` in the `layerStatesArray` for vectorlayers, but I don't think I have understood this part entirely.